[Machine-learning][ACR-1722][ACR-1723] Update ml-service documentation (#1697)
Samsung/tizen-docs: docs/application/native/guides/machine-learning/machine-learning-service.md

# Machine Learning Service

The Machine Learning Service API provides utility interfaces for both AI developers and application developers. AI developers can provide their pipelines as ML services, and application developers can use these services via the network.

The main features of the Machine Learning Service API include the following:

- Providing AI pipelines as ML services via machine-learning-agent (daemon).

  AI developers can manage their AI pipeline description with a unique name. Then they can launch the pipeline as an ML service.

- Using launched ML services via the network.

>   Application developers can request their data to be processed by the launched ML service. The communication is powered by [NNStreamer's tensor_query](https://nnstreamer.github.io/gst/nnstreamer/tensor_query/README.html) which implements [Edge-AI](https://nnstreamer.github.io/edge-ai.html) functionality.

Reviewer: ```suggestion
  Application developers can request their data to be processed by the launched ML service. The communication is powered by [NNStreamer's tensor_query](https://nnstreamer.github.io/gst/nnstreamer/tensor_query/README.html), which implements [Edge-AI](https://nnstreamer.github.io/edge-ai.html) functionality.
```

   <feature name="http://tizen.org/feature/machine_learning.service">true</feature>
   ```

## An example of object detection with Machine Learning Service API

> Let some AI developers want to provide a object detection service. The ML service takes an image data from other applications, and give back the image with bounding boxes drawn on it. An example pipeline can be described as this:

Reviewer: ```suggestion
Let's say some AI developers want to provide an object detection service. The ML service takes an image data from other applications, and give back the image with bounding boxes drawn on it. An example pipeline can be described as this:
```

```text
tensor_query_serversrc port=#reserved_appointed_port_number host=localhost connect-type=TCP !
other/tensors,num_tensors=1,types=uint8,format=static,dimensions=3:320:320:1,framerate=0/1 !
tensor_transform mode=arithmetic option=typecast:float32,add:0.0,div:255.0 !
queue ! tensor_filter framework=tensorflow2-lite model=/path/to/yolov5s/model/file.tflite !
other/tensors,num_tensors=1,types=float32,format=static,dimensions=85:6300:1:1,framerate=0/1 !
tensor_decoder mode=bounding_boxes option1=yolov5 option2=/path/to/coco/label.txt option4=320:320 option5=320:320 !
video/x-raw,format=RGBA,width=320,height=320 ! tensor_converter !
other/tensors,format=static,num_tensors=1,dimensions=4:320:320:1,types=uint8 !
tensor_query_serversink async=false sync=false
```

> The pipeline expects a 320x320 RGB raw image as input, and pre-process it to invoke the yolov5s tflite model. The inference result is decoded to bounding box image with tensor_decoder. The service provider can upgrade their internal neural network without changing the code of client if input and output layout is preserved.

Reviewer: ```suggestion
The pipeline expects a 320x320 RGB raw image as input, and pre-process it to invoke the yolov5s tflite model. The inference result is decoded to bounding box image with tensor_decoder. The service provider can upgrade their internal neural network without changing the code of the client if input and output layout is preserved.
```

> Clients who want to use this service may use their own pipeline communicating with the service, or can use Machine Learning Service API which is more easy to use.

Reviewer: ```suggestion
Clients who want to use this service may use their own pipeline communicating with the service, or can use the Machine Learning Service API, which is more easy to use.
```

The following code snippet shows an example of how the Machine Learning Service API can be used in this scenario. Note that checking return value and error handling is omitted for brevity.

### AI developers launching the object detection ML service

> AI developers can manage the pipeline description.

Reviewer: ```suggestion
AI developers can manage the pipeline description by doing the following:
```
Introduce a list with a complete sentence, not a partial one that's completed by the list items.
Reference: https://developers.google.com/style/lists#introductory-sentences-for-lists

    tensor_transform mode=arithmetic option=typecast:float32,add:0.0,div:255.0 ! \
    queue ! tensor_filter framework=tensorflow2-lite model=/path/to/yolov5s/model/file.tflite ! \
    other/tensors,num_tensors=1,types=float32,format=static,dimensions=85:6300:1:1,framerate=0/1 ! \
    tensor_decoder mode=bounding_boxes option1=yolov5 option2=/path/to/coco/label.txt option4=320:320 option5=320:320 ! \
    video/x-raw,format=RGBA,width=320,height=320 ! tensor_converter ! \
    other/tensors,format=static,num_tensors=1,dimensions=4:320:320:1,types=uint8 ! \
    tensor_query_serversink async=false sync=false";

    // Set the pipeline with service name.
    ml_service_set_pipeline ("object_detection_service", object_detection_pipeline);
    ```

2. Get or delete the pipeline description of the service name:

    ```c
    // Get the pipeline description of given service name.
    gchar *pipeline;
    ml_service_get_pipeline ("object_detection_service", &pipeline);

    // Delete the pipeline description of given service.
    ml_service_delete_pipeline ("object_detection_service");
    ```

> AI developers can manage the ML service.

Reviewer: ```suggestion
AI developers can manage the ML service by doing the following:
```
Introduce a list with a complete sentence, not a partial one that's completed by the list items.
Reference: https://developers.google.com/style/lists#introductory-sentences-for-lists

1. Launch and start the service:

    ```c
    // Launch the given service. Proper pipeline description is should be set beforehand.
    ml_service_h object_detection_service_h;
    ml_service_launch_pipeline ("object_detection_service", object_detection_service_h);

    // Start the given service. Clients can use this service when the pipeline is in playing state.
    ml_service_start_pipeline (object_detection_service_h);
    ```

2. Stop, destroy or get the pipeline state:

    ```c
    // Get the state of given service.
    ml_pipeline_state_e state;
    ml_service_get_pipeline_state (object_detection_service_h, &state);

    // Stop the given service.
    ml_service_stop_pipeline (object_detection_service_h);

    // Destroy the given service.
    ml_service_destroy (object_detection_service_h);
    ```

### Application developers using the service

> Clients (application developers) can use services with the Machine Learning Service API exploiting NNStreamer's tensor_query feature.

Reviewer: ```suggestion
Clients (application developers) can use services with the Machine Learning Service API exploiting NNStreamer's tensor_query feature. They can do this by using the following steps:
```
An ordered/unordered list should be preceded by a “colon”.
